File: 2023/day1/day1-part2.scala
package adventofcode2023.day1.part2
import scala.io.Source

val wordDigits = List(
  ("one", 1),
  ("two", 2),
  ("three", 3),
  ("four", 4),
  ("five", 5),
  ("six", 6),
  ("seven", 7),
  ("eight", 8),
  ("nine", 9),
)

case class WordMatch(word: String, digit: Int, index: Int)

def replaceLast(input: String, matchOn: String, replaceWith: String) = {
  input.reverse.replaceFirst(matchOn.reverse, replaceWith.reverse).reverse
}

def convertLineToDigits(line: String, direction: String) = {
  wordDigits
    .foldLeft(List[WordMatch]()){ case (acc, (word, digit)) =>
      val matches = word.r.findAllMatchIn(line)
      acc ++ matches.map(m => WordMatch(word, digit, m.start)).toList
    }
    .sortWith((a, b) => if (direction == "ltr") a.index.compareTo(b.index) < 0 else a.index.compareTo(b.index) > 0)
    .foldLeft(line){ case (acc, WordMatch(word, digit, index)) => 
      if (direction == "ltr") acc.replaceFirst(word, digit.toString) else replaceLast(acc, word, digit.toString)
    }
}

def findFirstDigit(charArray: Seq[Char]) = {
  charArray.find(_.isDigit).getOrElse('0').toString
}

def calculateLineCalibrationValue(line: String) = {
  (findFirstDigit(convertLineToDigits(line, "ltr")) +
      findFirstDigit(convertLineToDigits(line, "rtl").reverse)).toInt
}

def calculateCalibrationValueForFile(filename: String) = {
  val source = Source.fromFile(filename)
  source.getLines.foldLeft(0){ case (acc, line) =>
    // println(s"${line},${calculateLineCalibrationValue(line)}")
    // println(convertLineToDigits(line, "ltr"))
    acc + calculateLineCalibrationValue(line)
  }
}

@main def entryPoint() = {
  println(calculateCalibrationValueForFile("day1.txt"))
}

File: 2023/day1/day1-part2.mjs
import fs from 'fs'

const wordDigits = [
  ['one', 1],
  ['two', 2],
  ['three', 3],
  ['four', 4],
  ['five', 5],
  ['six', 6],
  ['seven', 7],
  ['eight', 8],
  ['nine', 9],
]

const convertLineToDigits = (line, wordToDigitInputs, direction) =>
  wordToDigitInputs
    .reduce((acc, [word, digit]) => {
      const matches = [...line.matchAll(word)]
      const found = matches.reduce((matchAcc, match) => [...matchAcc, { word, digit, index: match.index }], [])

      return [...acc, ...found]
    }, [])
    .sort((a, b) => (direction === 'ltr' ? a.index - b.index : b.index - a.index))
    .reduce(
      (acc, { word, digit }) =>
        acc.replace(direction === 'rtl' ? new RegExp(`(${word})(?!.*\\1)`) : word, digit.toString()),
      line
    )

const findFirstDigit = (charArray) => charArray.find((char) => !isNaN(parseInt(char)))

const calculateLineCalibrationValue = (line) =>
  parseInt(
    findFirstDigit([...convertLineToDigits(line, wordDigits, 'ltr')]) +
      findFirstDigit([...convertLineToDigits(line, wordDigits, 'rtl')].reverse())
  )

const calculateCalibrationValueForFile = (filename) =>
  fs
    .readFileSync(filename, 'utf-8')
    .split('\n')
    .reduce((acc, line) => {
      console.log(`${line},${calculateLineCalibrationValue(line)}`)
      return acc + calculateLineCalibrationValue(line)
    }, 0)

console.log(calculateCalibrationValueForFile('day1.txt'))
